docs/DEVELOPER_GUIDE.md

```diff
@@ -1 +1,179 @@
-todo
\ No newline at end of file
+# Development
+
+This doc explains how to setup a development environment so you can get started
+[contributing](../CONTRIBUTING.md)
+to `kfserving`. Also take a look at:
+
+- [How to add and run tests](../test/README.md)
+- [Iterating](#iterating)
+
+## Prerequisites
+
+Follow the instructions below to set up your development environment. Once you
+meet these requirements, you can make changes and
+[deploy your own version of kfserving](#deploy-kfserving)!
+
+Before submitting a PR, see also [CONTRIBUTING.md](../CONTRIBUTING.md).
+
+### Sign up for GitHub
+
+Start by creating [a GitHub account](https://github.com/join), then setup
+[GitHub access via SSH](https://help.github.com/articles/connecting-to-github-with-ssh/).
+
+### Install requirements
+
+You must install these tools:
+
+1. [`go`](https://golang.org/doc/install): KFServing controller is written in Go.
+1. [`git`](https://help.github.com/articles/set-up-git/): For source control.
+1. [`dep`](https://github.com/golang/dep): For managing external Go
+   dependencies.
+1. [`ko`](https://github.com/google/ko):
+   For development.
+1. [`kubectl`](https://kubernetes.io/docs/tasks/tools/install-kubectl/): For
+   managing development environments.
+1. [`kustomize`](https://github.com/kubernetes-sigs/kustomize/) To customize YAMLs for different environments
+
+### Install Knative on a Kubernetes cluster
+
+1. [Set up a kubernetes cluster and install Knative Serving](https://knative.dev/docs/install/)
+
+- **Note**: KFServing currently only requires `Knative Serving` for auto-scaling, canary rollout,
+ `Istio` for traffic routing and ingress. You can follow instructions on
+ [Custom Install](https://knative.dev/docs/install/knative-custom-install) to install `Istio` and `Knative Serving`,
+ observability plug-ins are good to have for monitoring.
+
+### Setup your environment
+
+To start your environment you'll need to set these environment variables (we
+recommend adding them to your `.bashrc`):
+
+1. `GOPATH`: If you don't have one, simply pick a directory and add
+   `export GOPATH=...`
+1. `$GOPATH/bin` on `PATH`: This is so that tooling installed via `go get` will
+   work properly.
+1. `KO_DOCKER_REPO`: The docker repository to which developer images should be
+   pushed (e.g. `docker.io/<username>/[project]`).
+
+- **Note**: Set up a docker repository for pushing images. You can use any container image registry by adjusting 
+the authentication methods and repository paths mentioned in the sections below.
+   - [Google Container Registry quickstart](https://cloud.google.com/container-registry/docs/pushing-and-pulling)
+   - [Docker Hub quickstart](https://docs.docker.com/docker-hub/)
+   - [Azure Container Registry quickstart](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-portal)
```

Comment:
quay.io is also a good registry https://quay.io/tutorial/

Reply:
Quay has had several large scale outages over the last year. My preference would be to use GCR, though perhaps we can maintain across multiple cloud vendors.

Reply:
this is for developer's testing image registry, three choices should be good enough :)

Reply:
For some developers in China, `quay.io` maybe a good choice. As Google Container Registry request `enable billing` and the `billing` now do not support Chinese credit card.

![image](https://user-images.githubusercontent.com/4461983/57264655-b16aec00-70a5-11e9-90c3-b8136837ada0.png)